unixfs/src/dir/builder/dir_builder.rs

```diff
@@ -0,0 +1,78 @@
+use super::{Entry, Leaf};
+use crate::Metadata;
+use std::collections::hash_map::Entry::*;
+use std::collections::HashMap;
+
+pub(super) struct DuplicateName;
+pub(super) struct FoundLeaf;
+
+/// Node in a directory tree.
+#[derive(Debug)]
+pub(super) struct DirBuilder {
+    /// Immediate files, symlinks or directories in this directory
+    pub nodes: HashMap<String, Entry>,
+    /// Metadata for this directory
+    metadata: Metadata,
+    /// Id of the parent; None for the root node
+    pub parent_id: Option<u64>,
+    /// Internal id, used for propagating Cids back from children during post order visit.
+    pub id: u64,
+}
+
+impl DirBuilder {
+    pub fn new(parent_id: u64, id: u64) -> Self {
+        assert_ne!(parent_id, id);
+        DirBuilder {
+            nodes: HashMap::new(),
+            metadata: Default::default(),
```

Comment:
`nodes` and `metadata` could be initialized with `..Default::default()` under the `id`

Reply:
Only via deriving `Default` which would allow production of invalid nodes (or just root nodes and entrenching root id == 0).